### ISSUE TITLE
tcmu_rbd_handle_timedout_cmd:669 rbd/rbd.vmware01: Timing out cmd.

### ISSUE THREAD

**Issue by Kaeltis** [open-iscsi/tcmu-runner#315]
During our nightly Ceph scrubbing window, our ESXi hosts regularily (almost every night) lose the iSCSI connection and basically lock up completely (hard reset needed).

Looking at the logs, I found the following in the tcmu-runner logs:

```
Oct 18 02:03:29 node1.ceph tcmu-runner[1650]: tcmu_rbd_handle_timedout_cmd:669 rbd/rbd.vmware01: Timing out cmd.
Oct 18 02:03:29 node1.ceph tcmu-runner[1650]: tcmu_notify_conn_lost:176 rbd/rbd.vmware01: Handler connection lost (lock state 1)
Oct 18 02:03:29 node1.ceph tcmu-runner[1650]: 2017-10-18 02:03:29.987 1650 [ERROR] tcmu_rbd_handle_timedout_cmd:669 rbd/rbd.vmware01: Timing out cmd.
Oct 18 02:03:29 node1.ceph tcmu-runner[1650]: 2017-10-18 02:03:29.987 1650 [ERROR] tcmu_notify_conn_lost:176 rbd/rbd.vmware01: Handler connection lost (lock state 1)
Oct 18 02:03:29 node1.ceph tcmu-runner[1650]: 2017-10-18 02:03:29.988 1650 [INFO] tgt_port_grp_recovery_thread_fn:253: Disabled iscsi/iqn.2017-01.com.domain.ceph:iscsigw/tpgt_1.
Oct 18 02:03:29 node1.ceph tcmu-runner[1650]: tgt_port_grp_recovery_thread_fn:253: Disabled iscsi/iqn.2017-01.com.domain.ceph:iscsigw/tpgt_1.
Oct 18 02:03:30 node1.ceph tcmu-runner[1650]: tgt_port_grp_recovery_thread_fn:278: Enabled iscsi/iqn.2017-01.com.domain.ceph:iscsigw/tpgt_1.
Oct 18 02:03:30 node1.ceph tcmu-runner[1650]: 2017-10-18 02:03:30.154 1650 [INFO] tgt_port_grp_recovery_thread_fn:278: Enabled iscsi/iqn.2017-01.com.domain.ceph:iscsigw/tpgt_1.
Oct 18 02:03:33 node1.ceph tcmu-runner[1650]: tcmu_rbd_lock:446 rbd/rbd.vmware01: Acquired exclusive lock.
Oct 18 02:03:33 node1.ceph tcmu-runner[1650]: 2017-10-18 02:03:33.841 1650 [WARN] tcmu_rbd_lock:446 rbd/rbd.vmware01: Acquired exclusive lock.
```

**Comment by mikechristie** [comment]
What version of ESX is this with and what version is the wmfs being used for the VMs?

Did you use the ceph-iscsi-* tools to setup the iscsi targets, and did you make the necessary changes to the ESX host described here http://docs.ceph.com/docs/master/rbd/iscsi-initiator-esx/ (there is just the one setting the iscsi RecoveryTimeout to set to 25 secs).

It looks like commands are taking longer than normal due to the scrub activity. On the iscsi target side it looks like we recovered quickly and ESX relogged in right away, so IO should have continued on this target or one of the other ones if it acted similarly.

Maybe we hit a ESX guest or host timer that set the device offline. Could you attach the ESX kernel log (/var/log/vmkernel.log) and the system log (/var/log/messages) for the iscsi gw where you see the above messages?

**Comment by Kaeltis** [comment]
I'm using 2 ESXi 6.5 Hosts controlled by a vCenter 6.5 Appliance with VMFS 6

I created the iscsi target on 3 of my ceph hosts with the ceph-iscsi-* tools.

Since I created the hosts a while back and the documentation didn't exist back then, I did **NOT** have the RecoveryTimeout set to 25, I changed that now on both hosts.

The documentation talks about setting the Path to "Most Recently Used (VMware)", however I can't find that anywhere, did they remove that in 6.5?

One thing I've noticed is that the performance of the ESXi Webinterface decreased extremely after adding the Ceph iSCSI Target, especially when clicking on "Storage" on the left side (I',m guessing every time it queries/reloads the Storage)

Since my scratch partition stopped working after enabling iSCSI (a known bug with the Dell customized image) I don't have any log from before the crash. I applied the workaround and will provide a log when it happens again.

**Comment by mikechristie** [comment]
> I'm using 2 ESXi 6.5 Hosts

Are you doing vsphere clustering? If so there is a bug in the ceph iscsi code where if you have a LUN that has VMs running on different hosts and a path to only one of them goes down then perf will be very bad because the backend iscsi target is going to be failing over in the background.

As a temp workaround you need to put VMs on different LUNs.

> The documentation talks about setting the Path to "Most Recently Used (VMware)", however I can't find that anywhere, did they remove that in 6.5?

I also have not seen it in the web interface for 6.5. MRU is the default so you should be ok. You can view/change it with the cmd line. ssh into the host and do:

// get devices.
esxcli storage  nmp device list 

then

// get multipath info for a device
esxcli storage nmp path list -d your_device_ide_from_list_command.

and it should show info including the path policy.

> I applied the workaround and will provide a log when it happens again.

The recovery timeout is not a workaround. It will just delay failover so the targets have time to recover.

If the host is locking up, I think you might be hitting a APD timeout issue. Commands might be slow through all paths during the scrub and so we get ceph iscsi rbd timeouts on all of them and all paths go down. The default timeout for that scenario is 140 seconds. We should not normally be hitting this plus the normal timeout for a command during scrub though.

Have you by any chance lowered the APD timeout:
https://docs.vmware.com/en/VMware-vSphere/6.5/com.vmware.vsphere.storage.doc/GUID-5CC5A990-AC23-4D51-80DD-37F09175E255.html

Send me the logs for when you hit this issue and I will check it out.

Also, if you can send me the target side and esx logs during the web interface slow down, I will look into that too.

**Comment by Kaeltis** [comment]
> Are you doing vsphere clustering?

Yes, I am. So for now I can not use clustering/HA and have to use different LUNs for different hosts instead of sharing one between all of them? Is there any estimate on when this will be fixed?

But - this does not explain the general slowness and the problems/lockup happening when the ceph cluster is just under scrubbing load, since no path is actually going down, right?

> esxcli storage nmp path list -d your_device_ide_from_list_command

This is the output from the path command:
```
iqn.1998-01.com.vmware:esx01-78928ec7-00023d000004,iqn.2017-01.com.domain.ceph:iscsigw,t,1-naa.60014059731a7108ea74de29e8c1c55f
   Runtime Name: vmhba64:C0:T0:L0
   Device: naa.60014059731a7108ea74de29e8c1c55f
   Device Display Name: LIO-ORG iSCSI Disk (naa.60014059731a7108ea74de29e8c1c55f)
   Group State: active
   Array Priority: 0
   Storage Array Type Path Config: {TPG_id=1,TPG_state=AO,RTP_id=1,RTP_health=UP}
   Path Selection Policy Path Config: {current path; rank: 0}

iqn.1998-01.com.vmware:esx01-78928ec7-00023d000006,iqn.2017-01.com.domain.ceph:iscsigw,t,3-naa.60014059731a7108ea74de29e8c1c55f
   Runtime Name: vmhba64:C2:T0:L0
   Device: naa.60014059731a7108ea74de29e8c1c55f
   Device Display Name: LIO-ORG iSCSI Disk (naa.60014059731a7108ea74de29e8c1c55f)
   Group State: active unoptimized
   Array Priority: 0
   Storage Array Type Path Config: {TPG_id=3,TPG_state=ANO,RTP_id=3,RTP_health=UP}
   Path Selection Policy Path Config: {non-current path; rank: 0}

iqn.1998-01.com.vmware:esx01-78928ec7-00023d000005,iqn.2017-01.com.domain.ceph:iscsigw,t,2-naa.60014059731a7108ea74de29e8c1c55f
   Runtime Name: vmhba64:C1:T0:L0
   Device: naa.60014059731a7108ea74de29e8c1c55f
   Device Display Name: LIO-ORG iSCSI Disk (naa.60014059731a7108ea74de29e8c1c55f)
   Group State: active unoptimized
   Array Priority: 0
   Storage Array Type Path Config: {TPG_id=2,TPG_state=ANO,RTP_id=2,RTP_health=UP}
   Path Selection Policy Path Config: {non-current path; rank: 0}
```

> The recovery timeout is not a workaround. It will just delay failover so the targets have time to recover.

with "workaround" i meant Dell's fix of the scratch partition, so the logs are persisted again, but I also applied the recovery timeout of 25.

> Have you by any chance lowered the APD timeout

No, it's set to 140 on both hosts.

> Also, if you can send me the target side and esx logs during the web interface slow down, I will look into that too.

I attached the logs, captured after clicking around, stopping and starting VMs, rescanning the storage devices, ... to cause some load. Hope it helps. I'll post a more detailed log tomorrow, after our nightly scrubbing time!

[vmkernel.log](https://github.com/open-iscsi/tcmu-runner/files/1396378/vmkernel.log)

The only thing I'm currently seeing on the target side is in dmesg:
```
[1382567.124995] ABORT_TASK: Found referenced iSCSI task_tag: 123417
[1382575.529050] ABORT_TASK: Sending TMR_FUNCTION_COMPLETE for ref_tag: 123417
[1382575.529056] ABORT_TASK: Found referenced iSCSI task_tag: 243245
[1382575.529057] ABORT_TASK: Sending TMR_TASK_DOES_NOT_EXIST for ref_tag: 243245
[1382575.530324] ABORT_TASK: Found referenced iSCSI task_tag: 243255
[1382575.532547] ABORT_TASK: Sending TMR_FUNCTION_COMPLETE for ref_tag: 243255
```

Thank you so much for your time and help!

**Comment by mikechristie** [comment]
>  Is there any estimate on when this will be fixed?

Yes, correct sharing the lun between multiple hosts is not supported right now.

It is the next item on the TODO after this release. Maybe a month.

> But - this does not explain the general slowness and the problems/lockup happening when the ceph cluster is just under scrubbing load, since no path is actually going down, right?

It might explain it. It does not explain why commands are taking so long initially, but once commands start taking a long time (longer than the rados osd op timeout of 30 seconds) paths will go down temporarily. So if commands started getting slow on all paths at the same time then all paths could go down temporarily. However, from the log snippet it looked like we were recovering really quickly so we should not have got into the situation where you had to hard reset. 

Here is what we are seeing from the target logs in the original message:

// This log message it means a command took longer than 30 seconds to complete. We did not think it would take more than 30 secs to complete a command even with scrub in progress, so the code assumes it might be a connection to the ceph cluster issue. We are going to fail the command and ask for a retry while we try to reconnect the rbd image to the ceph cluster. (Note that in your case this assumption might be wrong and for some reason it really is taking the OSD longer than 30 secs to complete the command).

Oct 18 02:03:29 node1.ceph tcmu-runner[1650]: tcmu_rbd_handle_timedout_cmd:669 rbd/rbd.vmware01: Timing out cmd.

// Here we disabled the iscsi target port while we try to reopen the image

Oct 18 02:03:29 node1.ceph tcmu-runner[1650]: tgt_port_grp_recovery_thread_fn:253: Disabled iscsi/iqn.2017-01.com.domain.ceph:iscsigw/tpgt_1.

// We reopened the image and are now renabling the iscsi target port. It looks like it was not a issue with the connection to the ceph cluster because we were able to reopen the image in 1 second.

Oct 18 02:03:30 node1.ceph tcmu-runner[1650]: tgt_port_grp_recovery_thread_fn:278: Enabled iscsi/iqn.2017-01.com.domain.ceph:iscsigw/tpgt_1.


// ESX relogged into the iscsi target port pretty quickly (3 seconds) and started to send IO to this path. This path grab the rbd lock so it can execute IO.

Oct 18 02:03:33 node1.ceph tcmu-runner[1650]: tcmu_rbd_lock:446 rbd/rbd.vmware01: Acquired exclusive lock.

> maybe the additional load by the flapping OSDs makes it even worse?

If OSDs are going down while a scrub is in progress then it could explain why commands are taking a long time. We might not be detecting this and failing over to other OSDs in time. On your OSDs in your /etc/ceph/ceph.conf  did you set these settings (it is mentioned in the doc I referenced so if you did not have the doc initially you missed it):

[osd]
osd heartbeat grace = 20
osd heartbeat interval = 5

?

> [1382567.124995] ABORT_TASK: Found referenced iSCSI task_tag: 123417
> [1382575.529050] ABORT_TASK: Sending TMR_FUNCTION_COMPLETE for ref_tag: 123417

These messages mean ESX's command timer has fired. It sent an abort, but the command completed shortly after that (so this just means nothing was permanently hung). Depending on the command type that could be 8 - 15 seconds. So during a scrub if IO is going to take a long time then these are expected. If OSDs are crashing we would also expect them. If you you just doing normal IO and/or just opening the storage part of the gui we should not see them. When we see these on the target side, we want to check if the ESX vmkernel log is mentioning something about APD and/or failovers. See below for info on how to increase the ESX logging for this and the scrub issue.

Oh yeah, for the path selection policy info it would just be in the 

esxcli storage  nmp device list 

command output. For your device naa.60014059731a7108ea74de29e8c1c55f you should see this for the type and policy:

   Storage Array Type: VMW_SATP_ALUA
   Path Selection Policy: VMW_PSP_MRU


> vmkernel.log

There was nothing interesting in here. There is one failed command during probing but that it is expected because we do not support it. ESX will ignore it.

Can you enable some extra logging. It will give more info about what is happening when we see those ABORT messages on the target.

In the Software Advanced Settings screen in the GUI under Scsi, set:

Scsi.LogCmdErrors 1
Scsi.LogScsiAborts 1
Scsi.LogMPCmdErrors 1
 
Could you also tell me the value for

Scsi.FailVMOnAPD

that you have.

> Thank you so much for your time and help!

No problem.

**Comment by Kaeltis** [comment]
Thanks for explaining everything, it really helps to understand the issue! 😄 

> did you set these settings

yes, I have:

osd client watch timeout = 15
osd heartbeat grace = 20
osd heartbeat interval = 5

the first line is from an older version of the docs, is this still needed?

> For your device naa.60014059731a7108ea74de29e8c1c55f you should see this for the type and policy

got that :)

> Can you enable some extra logging

just enabled it, I attached the logs from tonight, lockup happened again (seems to have stabilized now), many errors. There's also an NFS Storage attached, but it's only used for ISO images.

[logs.zip](https://github.com/open-iscsi/tcmu-runner/files/1402586/logs.zip)

FailVMIOonAPD is set to 0

**Comment by mikechristie** [comment]
> the first line is from an older version of the docs, is this still needed?

You do not need to change the watch timeout anymore. We have not udpated the docs. Do not worry about it. The other 2 settings are needed.

For the target side logs, could you attach the same file you saw the tcmu_rbd_handle_timedout_cmd log messages in? Was that /var/log/messages?

**Comment by Kaeltis** [comment]
Yes, I have the messages from around oct 20th attached:

[messages-20171022.zip](https://github.com/open-iscsi/tcmu-runner/files/1411902/messages-20171022.zip)

**Comment by mikechristie** [comment]
Just a FYI,

I am doing some experimenting here with different timers/settings which might work around the issue.

Will try to get back to you by Thurs/Fri.

**Comment by Kaeltis** [comment]
Hey @mikechristie, any news on the issue? Everything went relatively smooth the last couple of days (only problems during the nightly scrubbing time) but today one of the ESXi server "locked up" (webinterface shows root password wrong, so no login possible, restart via F12 on console hangs forever) during the scrubbing time again.

**Comment by mikechristie** [comment]
Hey,

Sorry. I am still testing.

Right now to work around the timeout issue I think we just need to increase these timers:

1. increase the rados osd op timeout (this is set for the iscsi gw in

/etc/ceph/iscsi-gateway.cfg

by adding

[target]
osd_op_timeout = N-in-seconds
 
and then restarting the gw.

2. decrease the osd heartbeat and interval settings like above.
3. increase the guest OS disk/command timeout per the guest OS docs.

It looks like a lot of times the first timer is just a little too short, so maybe setting it to around 60 - 90 seconds would be long enough so that we do not fire off the error handler when we do not need tol. #2 is probably ok for now, but dropping it down a little might help when we see the osd crashes. However, dropping too low is going to cause a lot of warnings in the logs, so I would leave it for now. And I am not yet sure about #3 when 1 - 2 are adjusted. #3 should be at least as long as #1 plus some settle/adjustment time. Because the entire host hung we never saw if the guest OS started to see errors too, so I would leave this for now.

I think we are also hitting something like this:

https://kb.vmware.com/s/article/2113956?language=en_US

There are so many ATS miscompare errors in the logs. We do not see the vmkernel errors mentioned  in the KB though, but I think it is worth trying. Note that for HA we do not support completely disabling ATS heartbeats because we do not support distributed reservations yet, so you should only disable ATS for heartbeats by doing (ignore the wmfs versions in the command):

esxcli system settings advanced set -i 0 -o /VMFS3/UseATSForHBOnVMFS5


Note that I am still testing this stuff though. You can try this out if this is not a production system.

**Comment by Kaeltis** [comment]
I'll hold on for now and leave everything as is (since I don't want to break anything), please let me know if you find someting out.
Thanks again for your help, I really appreciate it!

**Comment by Kaeltis** [comment]
Any news on this? My boss is breathing down my neck, since the 2 servers are sitting here, doing nothing 😆

**Comment by mikechristie** [comment]
We are working on it.

Jason is implementing offloaded ATS to avoid so many ATS errors:

https://github.com/open-iscsi/tcmu-runner/pull/330

and upstream we have changed our queueing:

https://www.spinics.net/lists/target-devel/msg16162.html

I think those combined with the updated settings will fix the issues. Those patches are being reviewed.

**Comment by Kaeltis** [comment]
Thanks again :) sorry for being such a nag, you guys are doing awesome work 👍

**Comment by vatelzh** [comment]
@mikechristie ,sorry for bother you.

> Are you doing vsphere clustering? If so **there is a bug in the ceph iscsi code** where if you have a LUN that has VMs running on different hosts and a path to only one of them goes down then perf will be very bad because the backend iscsi target is going to be failing over in the background.

Is this bug fixed?

**Comment by mikechristie** [comment]
No. For now, you would have to limit your setup to a VM per lun to work around the problem.

Just FYI, I do not think this is the issue you are hitting with gwcli. The problem here only affects the IO path and should not cause gwcli to be slow.

**Comment by vatelzh** [comment]
@mikechristie  I have a environment which has this scenario, VMs in different hosts use the same LUN, and the performance is very poor :(
gwcli is another performance issue.

**Comment by mikechristie** [comment]
>  @mikechristie I have a environment which has this scenario, VMs in different hosts use the same LUN, and the performance is very poor :(

1. You would only hit the performance issue I mentioned above if one host has failed the active-optimized path, but the other host can still access its active-optimized path. The problem is the rbd exclusive lock for that lun will bounce between the 2 gws.

2. If you are not hitting #1, you might be hitting a issue with ESX queue throttling and adaptive queueing. We are only starting to do performance work now. 2 things we know with ESX is that its the ESX SIOC adaptive queueing is causing the ESX device queue depth to be lowered too low. If you watch ESXTOP, you will see the DQLEN fluctuate. As a temp workaround you would have to disable this.

There is another issue where with the RHEL/Centos kernel the memory a tcmu devices uses is limited to 8 MB per device by default. Because the buffer used is small it can easily cause excessive QUEUE FULLs and the ESX queue full throttling might kick in.  For this you want to disable ESX queue full throttling and/or increase a setting on the ceph iscsi gw. Longer term we are trying to just kill that buffer/limit.

We are documenting this now and will update the ceph.com ESX info with the details.

In general performance is low right now. With nvme/ssds and 10 gig ethernet, throughput might be as low as 200-400 MB/s and latencies can be in the 50-70 ms range depending on IO sizes, workloads, etc.

**Comment by mikechristie** [comment]
Sorry forgot to update this issue. The last comments in it made me forget the root issue.

The needed pieces for CAW/ATS offloading are in ceph v13 and up and tcmu-runner 1.3.0. Note that tcmu-runner must be rebuilt against the updated librbd and updated librbd headers (so if you are using  prebuilt rpms make sure you also get the devel rpms) to take advantage of the feature.

Internally engineering has tested, but our QE team has not fully tested it yet.
